### ISSUE TITLE
[Proposition] Adopt JUnit 5 & AssertJ

### ISSUE THREAD

**Issue by echebbi** [gemoc/ale-lang#139]
## Proposal 

Tests are currently written with JUnit 4 and I believe we should move to JUnit 5:

 - JUnit 5 makes easier to parameterize a test 
 - JUnit 5 makes easier to organize tests (e.g. through `@Nested` and `@DisplayName`):

<div align = "center">

![image](https://user-images.githubusercontent.com/25926433/83396204-ea39d980-a3fb-11ea-8dd0-d0205759433f.png)

</div>

I believe this is important because some test classes are becoming heavy (`BuildTest` is 1k lines, `EvalTest` is 1.3k and `TypeValidatorTest` 1.6k) which makes harder and harder to browse existing tests and can lead to duplicated tests.

Moreover, I suggest adopting AssertJ which fluent API makes tests much more readable:

```java
@Test @DisplayName("executes all its tasks")
void executes_all_its_tasks() {
    split.run(context);
    assertThat(split.branches()).allMatch(Activity::hasRun);
}
```

AssertJ also provides [SoftAssertions](https://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/SoftAssertions.html), a way to test several predicates at once. We currently have a lot of tests that stop early in case of failure and do not check all their predicates which hence could benefit from this API, e.g.:

https://github.com/gemoc/ale-lang/blob/cfa611dd4deb0c0f36ffce38fd0682e531c6afb5/tests/org.eclipse.emf.ecoretools.ale.tests/src/org/eclipse/emf/ecoretools/ale/core/validation/test/TypeValidatorTest.java#L1841-L1847

## To be done

 - [ ] Add JUnit 5 & AssertJ to the target platform (warning: they are in Orbit, which may slow down first Maven build)
 - [ ] Make current tests use JUnit 5 and AssertJ

## Discussion

@dvojtise do you see any reason to stick to JUnit 4?

**Comment by dvojtise** [comment]
There is no particular reason to stick to JUnit 4

If someone migrates the test suites I'll be happy to merge :-)

**Comment by echebbi** [comment]
Great, I'll try to find some time then. By the way, we may have already discuss this but can we use snake_case to name JUnit tests? I usually try to use long, explicit and meaningful names which are hard to read with camel case:

```java
class ExecutionTest {

    @Test
    void when_launched_executes_its_tasks_in_sequential_order() {}

    @Test
    void while_running_when_canceled_stops_running() {}

    @Test
    void while_running_when_canceled_has_canceled_state() {}

}
```

Names are a bit shorter with JUnit 5, but I still find snake_case to be more readable (full example [here](https://github.com/echebbi/ekumi/blob/master/tests/fr.kazejiyu.ekumi.core.impl.test/src/fr/kazejiyu/ekumi/core/execution/impl/JobsExecutionTest.java)):

```java
class ExecutionTest {

    @Nested @DisplayName("when running")
    class WhenRunning {

        @BeforeEach
        void runExecution() {}

        @Nested @DisplayName("when canceled")
        class WhenCanceled {

            @BeforeEach
            void cancelExecution() {}

            @Test
            void stops_running() {}

            @Test
            void has_canceled_state() {}
        }
    }

}
```

**Comment by dvojtise** [comment]
you're right snake case looks nice for test names, I usually use camelCase mainly because I reuse the general rule for method names  but for tests, the snake case looks better.
